### PR TITLE
Set the output pipe as a non-block fd and catch EAGAIN when we write into it

### PR DESCRIPTION
**Pull request by dinosaure**
/cc @haesbaert this is my first try to fix the pipe used to interrupt a domain. WDYT?

### CODE REVIEW

File: lib/miou_unix.ml [hunk 1]
```diff
@@ -331,12 +331,11 @@ let transmit_fds signals revert tbl fds =
 
 let interrupted fd fds = List.exists (( = ) fd) fds
 
-let intr fd interrupted =
+let intr fd =
   let buf = Bytes.create 0x100 in
   ignore (Unix.read fd buf 0 (Bytes.length buf));
```
Comment (haesbaert) [inline lib/miou_unix.ml:336]: I'd also suggest to pre-allocate buf with a longer lenght.

File: lib/miou_unix.ml [hunk 2]
```diff
@@ -367,25 +366,25 @@ let select uid (interrupt, p) ~block cancelled_syscalls =
...
-    else ignore (Atomic.fetch_and_add interrupted 1)
+let interrupt oc () =
+  match Unix.single_write oc signal 0 1 with
+  | _ -> ()
+  | exception Unix.(Unix_error (EAGAIN, _, _)) -> ()
 
 and signal = Bytes.make 1 '\000'
 
 let events domain =
   let ic, oc = Unix.pipe ~cloexec:true () in
```
Comment (haesbaert) [inline lib/miou_unix.ml:384]: You can also make ic non-blocking, and handle EAGAIN on read as well, on the odd chance that something is slightly desynced, you prevent an erroneous block on read.

File: lib/miou_unix.ml [hunk 3]
```diff
@@ -367,25 +366,25 @@ let select uid (interrupt, p) ~block cancelled_syscalls =
...
       let signals = transmit_fds signals domain.revert domain.writers wrs in
       signals
 
-let rec interrupt oc interrupted () =
-  if Atomic.get interrupted = 0 then
-    if Unix.single_write oc signal 0 1 = 0 then interrupt oc interrupted ()
-    else ignore (Atomic.fetch_and_add interrupted 1)
+let interrupt oc () =
+  match Unix.single_write oc signal 0 1 with
+  | _ -> ()
```
Comment (haesbaert) [inline lib/miou_unix.ml:378]: I'd test that I've actually written one byte, more for pedantry, as it's actually impossible due to the underlying implementation, but the basically writes on stream can/could always return shortcounts.

**Review by haesbaert — APPROVED** [review]
Looks good, I've added some additional suggestions, I think the non-blocking read is worth having as well, the rest is because haesbaert is a pedantic fanatical :-).

File: lib/miou_unix.ml [hunk 4]
```diff
@@ -330,10 +330,11 @@ let transmit_fds signals revert tbl fds =
   List.fold_left fold signals fds
 
 let interrupted fd fds = List.exists (( = ) fd) fds
+let buf = Bytes.create 0x1000
```
[the remark this answers is not in the document]
Comment (dinosaure) [reply lib/miou_unix.ml:333]: Yes, I was too fast about my last commit, I will fix that and revert as before. Thrme allocation of a small one does not cost too much.

**Pull request merged**